docs/RFCS/leader_evaluated_raft.md

```diff
@@ -0,0 +1,425 @@
+before proposing the command to Raft, much of this downstream code moves
+upstream of Raft; naively what's left are simple writes to key-value pairs
+(i.e. all commands move upstream except for a simplified MVCCPut operation).
+While some complexity remains, this has promise for a much facilitated
+migration story, as well as de-triplicating work previously done by each
+individual replica.
+
+# Detailed design
+
+## Execution path
+
+### Proposing commands
+
+The main change in the execution path will be in `proposeRaftCommand`, which
+currently takes a `BatchRequest` and creates a `pendingCmd`, which currently
+has the form
+
+```go
+type pendingCmd struct {
+  ctx     context.Context
+  idKey   storagebase.CmdIDKey
+  done    chan roachpb.ResponseWithError // Used to signal waiting RPC handler
+  raftCmd struct {
+      RangeID       RangeID
+      OriginReplica ReplicaDescriptor
+      Cmd           BatchRequest
+  }
+}
+```
+
+and within which the main change is changing the type of `Cmd` from
+`BatchRequest` to a new type which carries the effects of the application of
+the batch, as computed (by the leader) earlier in `proposeRaftCommand`.
+
+For simplicity, in this draft we will assume that the `raftCmd` struct will
+be modified according to these requirements.
+
+The "effects" contain
+
+* any key-value writes: For example the outcome of a `Put`, `DeleteRange`, `CPut`, `BeginTransaction`, or `LeaderLease`.
+  Some alternatives exist: writes could be encoded as `[]MVCCKeyValue`, or even
+  as post-MVCC raw (i.e. opaque) key-value pairs. We'll assume the former for
+  now.
+* parameters for any "triggers" which must run with the *application* of the
+  command. In particular, this is necessary on `EndTransaction` with a commit
+  trigger, or when leadership changes (to apply the new lease).
+  There are more of them, but this already shows that these triggers (which are
+  downstream of Raft) will require various pieces of data to be passed around
+  with them.
+
+Note that for the "triggers" above, a lot of the work they currently do can
+move pre-Raft as well. For example, for `LeaderLease` below we must change some
+in-memory state on `*Replica` (lease proto and timestamp cache) plus call out
+to Gossip. Still, the bulk of the code in `(*Replica).LeaderLease` (i.e. the
+actual logic) will move pre-Raft, and the remaining trigger is conditional on
+being on the leader (so that one could even remove it completely from the
+"effects" and hack their execution into the leader role itself).
+
+Thus, we arrive at the following (proto-backed) candidate struct:
+
+```proto
+# Pseudo-Go-Protobuf, all naming ad-hoc and TBD.
+
+message RaftCmd {
+  optional SomeMetadata # tbd (proposing replica, MVCCStats diff, ...)
+  repeated oneof effect {
+    # Applied in field and slice order.
+    optional Writes writes = ...; # see MVCC section for Writes type
+    # Carries out the following:
```

Comment:
In the proposed RaftCmd, I assume you're just going to use the various Trigger protobufs we already have? Let's not create new types for this struct.